packages/Search/src/details/DTK_DetailsTreeTraversal.hpp

```diff
@@ -42,242 +42,208 @@ struct TreeTraversal
         return queryDispatch( Tag{}, bvh, pred, std::forward<Args>( args )... );
     }
 
-    /**
-     * Return true if the node is a leaf.
-     */
-    KOKKOS_INLINE_FUNCTION
-    static bool isLeaf( Node const *node )
-    {
-        return ( node->children.first == nullptr );
-    }
-
-    /**
-     * Return the index of the leaf node.
-     */
-    KOKKOS_INLINE_FUNCTION
-    static size_t getIndex( Node const *leaf )
-    {
-        static_assert( sizeof( size_t ) == sizeof( Node * ),
-                       "Conversion is a bad idea if these sizes do not match" );
-        return reinterpret_cast<size_t>( leaf->children.second );
-    }
-
-    /**
-     * Return the root node of the BVH.
-     */
-    KOKKOS_INLINE_FUNCTION
-    static Node const *getRoot( BoundingVolumeHierarchy<DeviceType> const &bvh )
+    // There are two (related) families of search: one using a spatial predicate
+    // and one using nearest neighbours query (see boost::geometry::queries
+    // documentation).
+    template <typename Predicate, typename Insert>
+    KOKKOS_FUNCTION static int
+    spatialQuery( BoundingVolumeHierarchy<DeviceType> const &bvh,
+                  Predicate const &predicate, Insert const &insert )
```

Comment:
Github has trouble showing the diff in this file. Can you tell us what has changed.

Reply:
Yes I know it's not pretty, I gave up trying to cleanup history (about two dozens of commits where I was not able to use clang-format...)

Here is summary of what happened:
* Removed `TreeTraversal::isLeaf()`, `:: getIndex()`, and `::getRoot()` in favor of `bvh.isLeaf()`, `bvh.getLeafPermutationIndex()`, and `bvh.getRoot()`.
* Replaced `predicate( node )` and `distance( node )` by `predicate( bvh.getBoundingVolume( node ) )` and `distance( bvh.getBoundingVolume( node ) )`
* Moved `spatialQuery()` and `nearestQuery()` into `TreeTraversal` to grant them access to `BoundingVolumeHierarchy` private function members `getRoot()`, `getBoundingVolume()`, etc.